nova/tests/unit/test_utils.py

```diff
@@ -141,11 +142,40 @@ def test_hostname_truncated_no_hyphen(self):
 
     def test_generate_password(self):
         password = utils.generate_password()
-        self.assertTrue([c for c in password if c in '0123456789'])
-        self.assertTrue([c for c in password
-                         if c in 'abcdefghijklmnopqrstuvwxyz'])
-        self.assertTrue([c for c in password
-                         if c in 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'])
+        for group in utils.DEFAULT_PASSWORD_SYMBOLS:
+            if group:
+                self.assertTrue([c for c in password if c in group])
+
+    @mock.patch.object(CONF, "password_all_group_samples", 0)
+    def test_generate_password_zero_groups(self):
+        password = utils.generate_password()
+        self.assertEqual(len(password), CONF.password_length)
+
+    @mock.patch.object(CONF, "password_all_group_samples", 2)
+    def test_generate_password_two_groups(self):
+        password = utils.generate_password()
+        self.assertEqual(len(password), CONF.password_length)
+        for group in utils.DEFAULT_PASSWORD_SYMBOLS:
+            if group:
+                self.assertGreaterEqual(
+                    len([c for c in password if c in group]),
+                    CONF.password_all_group_samples)
+
+    @mock.patch.object(CONF, "password_all_group_samples", 7)
+    @mock.patch.object(CONF, "password_length", 13)
+    def test_generate_password_too_many_groups(self):
+        password = utils.generate_password()
+        self.assertEqual(len(password), CONF.password_length)
+
+    @mock.patch.object(CONF, "password_additional_symbols", string.punctuation)
+    def test_generate_password_additional_symbols(self):
+        password = utils.generate_password()
+        self.assertEqual(len(password), CONF.password_length)
+        for group in utils.DEFAULT_PASSWORD_SYMBOLS:
+            if group:
+                self.assertGreaterEqual(
+                    len([c for c in password if c in group]),
+                    CONF.password_all_group_samples)
```

Comment:
Imho this test doesn't really show/test that we have characters from `string.punctuation` in the password now. Instead, that test would also work if `CONF.password_additional_symbols` would not be used in the code at all.

I'd suggest to add a line `self.assertGreaterEqual(len([c for c in password if c in string.punctuation]), CONF.password_all_group_samples)`

Reply:
I've replaced the line with your proposal. Iterating over the groups is already done in another test, so it is kind of redundant.